### PR TITLE
use non-random images for interpolation kernels for testing

### LINKED ISSUE #6937: Check tolerances for prototype transforms kernel reference tests
Restoring the tolerances to

https://github.com/pytorch/vision/blob/f32600b66d9c03682ccd975e442e67c609fd7f0a/test/prototype_transforms_kernel_infos.py#L64-L67

that will be cranked up by #6934 surfaces failures on the following kernels in the reference test:

- [x] `adjust_brightness_image_tensor`
- [x] `adjust_contrast_image_tensor`
- [x] `adjust_gamma_image_tensor`
- [ ] `adjust_hue_image_tensor`
- [x] `adjust_saturation_image_tensor`
- [x] `adjust_sharpness_image_tensor`
- [x] `affine_image_tensor`
- [x] `affine_mask`
- [x] `autocontrast_image_tensor`
- [x] `center_crop_image_tensor`
- [x] `crop_image_tensor`
- [ ] `elastic_image_tensor`
- [ ] `elastic_mask`
- [x] `equalize_image_tensor`
- [x] `five_crop_image_tensor`
- [x] `horizontal_flip_image_tensor`
- [x] `invert_image_tensor`
- [x] `pad_image_tensor`
- [ ] `perspective_image_tensor`
- [x] `perspective_mask`
- [x] `posterize_image_tensor`
- [ ] `resize_image_tensor`
- [x] `resize_mask`
- [ ] `resized_crop_image_tensor`
- [x] `resized_crop_mask`
- [ ] `rotate_image_tensor`
- [x] `rotate_mask`
- [x] `solarize_image_tensor`
- [x] `ten_crop_image_tensor`
- [x] `vertical_flip_image_tensor`

We need to go through each of them to see if there is a systematic problem here. However, this is not reason to "panic" since

1. these kernels are used in the training pipeline and we are seeing no accuracy drop compared to v1
2. these kernels are also tested for consistency against v1 and there everything is fine.

cc @vfdev-5 @datumbox @bjuncek

### PR DESCRIPTION
**Pull request by pmeier**
Fixes #6937 by using option 1. proposed in https://github.com/pytorch/vision/issues/6937#issuecomment-1326236301.

cc @vfdev-5 @datumbox @bjuncek

### CODE REVIEW

File: test/prototype_common_utils.py [hunk 1]
```diff
@@ -69,6 +70,7 @@ def compare(self) -> None:
         self._compare_attributes(actual, expected)
 
         actual, expected = self._equalize_attributes(actual, expected)
+        actual, expected = self._promote_for_comparison(actual, expected)
```
Comment (pmeier) [inline test/prototype_common_utils.py:73]: This was a bug. Without this promotion, we are later subtracting `uint8` values for the absolute difference. They wrap around:

```pycon
>>> a = torch.tensor(10, dtype=torch.uint8)
>>> b = torch.tensor(11, dtype=torch.uint8)
>>> a - b
tensor(255, dtype=torch.uint8)
>>> abs(a - b)
tensor(255, dtype=torch.uint8)
```

If we convert to a singed integer dtype first (here `int64`) we get the correct absolute diff:

```pycon
>>> a = torch.tensor(10, dtype=torch.int64)
>>> b = torch.tensor(11, dtype=torch.int64)
>>> a - b
tensor(-1, dtype=torch.int64)
>>> abs(a - b)
tensor(1, dtype=torch.int64)
```

This bug only affected the comparisons that used `agg_method=...`.

File: test/prototype_common_utils.py [hunk 2]
```diff
@@ -313,6 +315,38 @@ def make_image_loaders(
...
+        num_channels, height, width = shape
+
+        top = int((0.8 * height) // 2)
+        left = int((0.8 * width) // 2)
+        bottom = height - top
+        right = width - left
+
+        image_pil = PIL.Image.new("L", (width, height))
+        draw = PIL.ImageDraw.Draw(image_pil)
+        draw.ellipse(((left, top), (right, bottom)), fill=255)
```
Comment (pmeier) [inline test/prototype_common_utils.py:332]: This draws a white ellipse onto black background with roughly 10% empty space on each side of the border. This way we get a non-trivial shape like a rectangle, but still have a large interior region. Assuming that the only differences happen at he border, using the MAE, i.e. `agg_method="mean"`, for comparison, leads to small tolerances and thus more safety.

File: test/prototype_transforms_kernel_infos.py [hunk 3]
```diff
@@ -319,17 +320,6 @@ def sample_inputs_resize_mask():
         yield ArgsKwargs(mask_loader, size=[min(mask_loader.shape[-2:]) + 1])
 
 
-@pil_reference_wrapper
-def reference_resize_mask(*args, **kwargs):
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:323]: Since all of our mask kernels are just wrapping the image ones, there is no need to compare both against a reference. Checking the image kernel is sufficient.

File: test/prototype_common_utils.py [hunk 4]
```diff
@@ -313,6 +315,38 @@ def make_image_loaders(
...
+
+        return features.Image(to_image_tensor(image_pil).repeat(num_channels, 1, 1), color_space=color_space).to(
+            dtype=dtype, device=device
+        )
+
+    return ImageLoader(fn, shape=(num_channels, *size), dtype=dtype, color_space=color_space)
+
+
+def make_image_loaders_for_interpolation(
+    sizes=((233, 147),),
```
Comment (pmeier) [inline test/prototype_common_utils.py:342]: I didn't use the default shapes here since they are pretty small and thus the area of the border of the ellipse is quite large compared to the overall image and thus larger tolerances would need to be larger. Open for discussion about this.

File: test/prototype_transforms_kernel_infos.py [hunk 5]
```diff
@@ -914,8 +869,7 @@ def sample_inputs_rotate_video():
             reference_fn=pil_reference_wrapper(F.rotate_image_pil),
             reference_inputs_fn=reference_inputs_rotate_image_tensor,
             float32_vs_uint8=True,
-            # TODO: investigate
-            closeness_kwargs=pil_reference_pixel_difference(110, agg_method="mean"),
+            closeness_kwargs=pil_reference_pixel_difference(0, agg_method="mean"),
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:872]: Oops, even better:

```suggestion
```

File: test/prototype_transforms_kernel_infos.py [hunk 6]
```diff
@@ -1426,11 +1364,12 @@ def sample_inputs_elastic_video():
...
             F.elastic_image_tensor,
             sample_inputs_fn=sample_inputs_elastic_image_tensor,
-            reference_fn=pil_reference_wrapper(F.elastic_image_pil),
             reference_inputs_fn=reference_inputs_elastic_image_tensor,
             float32_vs_uint8=float32_vs_uint8_fill_adapter,
-            # TODO: investigate
-            closeness_kwargs=float32_vs_uint8_pixel_difference(60, agg_method="mean"),
+            closeness_kwargs={
+                **float32_vs_uint8_pixel_difference(6, agg_method="mean"),
+                **cuda_vs_cpu_pixel_difference(),
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:1371]: Driveby that seems to have snuck in since the last time I was testing against CUDA.

File: test/prototype_transforms_kernel_infos.py [hunk 7]
```diff
@@ -1439,15 +1378,11 @@ def sample_inputs_elastic_video():
...
-            reference_fn=pil_reference_wrapper(F.elastic_image_pil),
-            reference_inputs_fn=reference_inputs_elastic_mask,
-            float32_vs_uint8=True,
-            # TODO: investigate
-            closeness_kwargs=pil_reference_pixel_difference(80, agg_method="mean"),
         ),
         KernelInfo(
             F.elastic_video,
             sample_inputs_fn=sample_inputs_elastic_video,
+            closeness_kwargs=cuda_vs_cpu_pixel_difference(),
```
Comment (pmeier) [inline test/prototype_transforms_kernel_infos.py:1385]: Same as above.

**Review by datumbox — APPROVED** [review]
LGTM. IMO the proposed approach is the better alternative from the other 4 described at https://github.com/pytorch/vision/issues/6937#issuecomment-1326236301

File: test/prototype_common_utils.py (hunk 2 above)
Comment (pmeier) [reply test/prototype_common_utils.py]: As discussed offline, instead of using this artificial image, we should use a real one. 2eba005893d8eeeb8f9cc2fd290e6e11cc6538d5 uses Grace Hopper.
